internal/flow/describe_release.go

```diff
@@ -0,0 +1,119 @@
+	"context"
+	"fmt"
+	"regexp"
+	"time"
+
+	"github.com/lunarway/release-manager/internal/artifact"
+	"github.com/lunarway/release-manager/internal/commitinfo"
+	"github.com/lunarway/release-manager/internal/git"
+	"github.com/lunarway/release-manager/internal/intent"
+	"github.com/lunarway/release-manager/internal/log"
+	"github.com/pkg/errors"
+)
+
+type DescribeReleaseResponse struct {
+	Releases []DescribeReleaseResponseRelease
+}
+
+type DescribeReleaseResponseRelease struct {
+	DefaultNamespaces bool
+	ReleaseIndex      int
+	Artifact          artifact.Spec
+	ReleasedAt        time.Time
+	ReleasedByEmail   string
+	ReleasedByName    string
+	Intent            intent.Intent
+}
+
+// DescribeRelease returns information about a specific release in an environment.
+func (s *Service) DescribeRelease(ctx context.Context, environment, service string, count int) (DescribeReleaseResponse, error) {
+	span, ctx := s.Tracer.FromCtx(ctx, "flow.DescribeRelease")
+	span.SetBaggageItem("count", fmt.Sprintf("%v", count))
+
+	defer span.Finish()
+	sourceConfigRepoPath, close, err := git.TempDirAsync(ctx, s.Tracer, "k8s-config-describe-release")
+	if err != nil {
+		return DescribeReleaseResponse{}, err
+	}
+	defer close(ctx)
+
+	log.WithContext(ctx).Debugf("Cloning source config repo %s into %s", s.Git.ConfigRepoURL, sourceConfigRepoPath)
+	sourceRepo, err := s.Git.Clone(ctx, sourceConfigRepoPath)
+	if err != nil {
+		return DescribeReleaseResponse{}, errors.WithMessagef(err, "clone into '%s'", sourceConfigRepoPath)
+	}
+
+	var releases []DescribeReleaseResponseRelease
+
+	var currentOffset uint = 0
+	for count-int(currentOffset) > 0 {
+		hash, err := s.Git.LocateServiceReleaseRollbackSkip(ctx, sourceRepo, environment, service, currentOffset)
+		if err != nil {
+			if errors.Is(err, git.ErrReleaseNotFound) {
+				break
+			}
+			return DescribeReleaseResponse{}, errors.WithMessagef(err, "locate release")
+		}
+
+		commitObj, err := sourceRepo.CommitObject(hash)
+		if err != nil {
+			return DescribeReleaseResponse{}, errors.WithMessagef(err, "get commit at hash '%s'", hash)
+		}
+
+		commitInfo, err := commitinfo.ParseCommitInfo(commitObj.Message)
+		if err != nil {
+			return DescribeReleaseResponse{}, errors.WithMessagef(err, "parse commit info at hash '%s'", hash)
+		}
+
+		currentNamespace := ""
+
+		r := regexp.MustCompile(fmt.Sprintf(`^(?P<environment>[^/]+)/releases/(?P<namespace>[^/]+)/(?P<service>[^/]+)/%s$`, regexp.QuoteMeta(s.ArtifactFileName)))
+
+		gitFilesSpan, _ := s.Tracer.FromCtx(ctx, "finding namespace in git commit stats")
+		gitFilesSpan.SetTag("gitcommit", commitObj.Hash.String())
+		stats, err := commitObj.Stats()
+		if err != nil {
+			return DescribeReleaseResponse{}, errors.WithMessagef(err, "could not find commit stats for %s", commitObj.Hash.String())
+		}
+		for _, stat := range stats {
+			match := r.FindStringSubmatch(stat.Name)
+			if match != nil {
+				currentNamespace = match[2]
+				break
+			}
+		}
+		gitFilesSpan.Finish()
```

Comment:
Could we extract this into a dedicated function? Just to simplify the method a bit?

Reply:
And further move `gitFilesSpan.Finish()` up to before the error handling?
In case of an error from `commitObj.Stats()` the span is never finished.

Reply:
Yeah, can do that 👍  by moving it into a function it also makes the span handling easier